Make clang-tidy check only the files it needs to

jbytheway opened the pull request:
#### Summary
Infrastructure "Make clang-tidy check only the files it needs to"

#### Purpose of change
clang-tidy is our slowest CI job.  It would be nice to be able to speed it up.  It's already only run when certain files are changes (e.g. not for json-only changes).

However, we can do more.  For small C++ changes that only affect `.cpp` files we should only need to run it on those specific files.  If a header or CMake file is altered, or a new check is added to the plugin, we need to run on everything.

#### Describe the solution

- [x] Figure out how to get a list of changed files on a GitHub PR.
- [x] Make sure the existing code to prioritize changed files is working.
- [x] Limit the files checked when possible.

#### Describe alternatives you've considered
None

#### Testing
Will require testing here on CI.  So, opening as draft initially.

#### Additional context
If this works out, I'd like to see this check made mandatory for merges.

kevingranade commented:
> If this works out, I'd like to see this check made mandatory for merges.

If a typical run is under an hour, it will be.

jbytheway commented:
> If a typical run is under an hour, it will be.

I think most PRs are still going to edit a header file, so they'd still be in the 2+hour range.  But it looks like it should be ~15 minutes for a PR only modifying a small number of cpp files.

kevingranade commented:
> I think most PRs are still going to edit a header file

More incentive to decouple those headers more thoroughly.

jbytheway commented:
> More incentive to decouple those headers more thoroughly.

For the avoidance of doubt, the current logic is: If any header changed then clang-tidy everything.  So decoupling headers won't help unless that allows you to move code to cpp files.

Making it check only files that rely on changed headers is possible, but that would require us to run the preprocessor on everything (note that we don't actually build the code in this CI job, so we don't have the `.d` files normally created by `make`).  That might be worthwhile, but out of scope for this PR.

kevingranade commented:
Ok so the current logic is, if any header or build files were touched, do a full check, but do the changed .cpp files first.
If only cpp files were touched, just check those files.

Anything left to check? The last two commits and workflow runs seem to be exercising that logic.

I agree only checking files with a changed header is out of scope, though I think it will be pretty easy.
From the GCC man page.
```
 -M  Instead of outputting the result of preprocessing, output a rule suitable for make describing the dependencies of the main source file.  The preprocessor outputs one make rule containing the object file name for that source file, a colon, and the names of all the included files
-MM Like -M but do not mention header files that are found in system header directories, nor header files that are included, directly or indirectly, from such a header.
```

jbytheway commented:
> Ok so the current logic is, if any header or build files were touched, do a full check, but do the changed .cpp files first.
> If only cpp files were touched, just check those files.

Correct.

> Anything left to check? The last two commits and workflow runs seem to be exercising that logic.

I think it's working now, so I'm marking ready for review.  I've removed the temporary changes and squashed everything down to one commit with a reasonable commit message.

> I agree only checking files with a changed header is out of scope, though I think it will be pretty easy.
> From the GCC man page.
<snip>

Indeed, that option exists, but to avoid surprises we probably need to ensure we're running the preprocessor with the same flags as clang-tidy will be (wrt `TILES`, optimization level, etc.).  We might be able to use the compilation database for that, the same as `clang-tidy` is.  But that is probably messy enough to warrant a Python script; not sure I'd want to do it in bash.

kevingranade commented:
> Indeed, that option exists, but to avoid surprises we probably need to ensure we're running the preprocessor with the same flags as clang-tidy will be (wrt TILES, optimization level, etc.)

Aha, good point I was just focusing on the core feature, but that stuff gets messy too.
